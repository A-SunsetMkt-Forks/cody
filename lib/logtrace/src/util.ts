import type { JsonObject, JsonValue, Jsonifiable } from 'type-fest'
import type { KeysOfUnion } from 'type-fest'
import type { UndefinedToOptional } from 'type-fest/source/internal'
import type { JsonifyObject } from 'type-fest/source/jsonify'
import { monotonicFactory } from 'ulidx'

export const IS_TEST = (typeof process === 'object' && !!process.env.VITEST) ?? false

export const BASE_PATH = (() => {
    if (typeof process !== 'undefined' && process.cwd) {
        // Node.js environment
        return process.cwd()
    }
    if (globalThis.location) {
        // Browser environment
        return globalThis.location.origin
    }
    return ''
})()

class IdGenenrator {
    private _instance = monotonicFactory()
    private _seedTime: number | undefined = undefined

    public setGenerator(
        prng: undefined | (() => number) = undefined,
        seedTime: number | undefined = undefined,
        force = IS_TEST
    ) {
        if (!force) {
            throw new Error(
                'IdGenerator should not be modified outside of test environments or without `force` flag'
            )
        }
        this._seedTime = seedTime
        this._instance = monotonicFactory(prng)
    }

    public next(): string {
        return this._instance(this._seedTime)
    }
}

function defaultDateGenFn() {
    return new Date()
}
class DateGenerator {
    private _instance = defaultDateGenFn
    private _timezoneOffset = defaultDateGenFn().getTimezoneOffset()
    // Useful for testing
    public setDateFn(
        fn: () => Date = defaultDateGenFn,
        timezoneOffset: number | undefined = undefined,
        force = IS_TEST
    ) {
        if (!force) {
            throw new Error(
                'DateGenerator should not be modified outside of test environments or without `force` flag'
            )
        }
        this._instance = fn
        this._timezoneOffset = timezoneOffset ?? fn().getTimezoneOffset()
    }

    public now() {
        return this._instance()
    }

    public get timezoneOffset() {
        return this._timezoneOffset
    }
}

export const dateGenerator = new DateGenerator()
export const idGenerator = new IdGenenrator()

// #region JSON
export type JsonifiableObject = {
    [Key in string]?:
        | Jsonifiable
        | { toJSON: () => Jsonifiable }
        | Array<{ toJSON: () => Jsonifiable }>
        | JsonifiableObject
}

type AllVariantFields<T> = {
    [K in KeysOfUnion<T>]?: never
}

// This field makes sure all fields in the union have a default value of never set.
// This makes it easy to use any field as a discriminant for the union
export type DefaultDiscriminantFields<T extends { _type: string }> = {
    [K in T as K['_type']]: K & Omit<AllVariantFields<T>, keyof K>
}[T['_type']]

// #region Possible Paths
//this type extracts all variants of objects that a key can be. It also takes any variants nested in array types
type UnionKeys<T> = T extends T ? (keyof T extends string | number ? keyof T : never) : never
type AggregateTypes<T, K extends PropertyKey> = T extends any
    ? K extends keyof T
        ? T[K]
        : never
    : never

type AggregatedObjectKeyType<T, K extends PropertyKey> = Extract<AggregateTypes<T, K>, object>
type AggregatedArrayKeyType<T, K extends PropertyKey> = Extract<
    AggregateTypes<T, K>,
    Array<any>
> extends Array<infer AV>
    ? Extract<AV, object>
    : never

/**
 * This bit of typing magic extracts all possible object paths from a heterogenous object.
 * This works even if the object has keys that are of complex union or array types.
 * These keys can be used with the @link{replacePathsIfSet} to substitute values.
 */
export type AllPossiblePaths<T> =
    | `${UnionKeys<T>}`
    | (Extract<T, object> extends any
          ? {
                  [K in UnionKeys<T>]:
                      | `${K}.${AllPossiblePaths<AggregatedObjectKeyType<T, K>>}`
                      | (Extract<T[K], Array<any>> extends Array<infer AV>
                              ? `${K}.[].${AllPossiblePaths<AggregatedArrayKeyType<T, K>>}`
                              : never)
              }[UnionKeys<T>]
          : never)

/**
 * Replace existing values, if they are not undefined/null, at the specified paths with the replacement value.
 * Paths are in the format as allowed by @link{AllPossiblePaths} such as `a.b.[].c`
 */
export function replacePathsIfSet<T extends object>(
    obj: JsonifyObject<UndefinedToOptional<T>>,
    paths: string[],
    replacement: JsonValue | JsonObject
) {
    for (const path of paths) {
        const [head, ...tail] = path.split('.')
        if (!obj || !(head in obj)) {
            continue
        }
        const valueAtKey = (obj as any)?.[head]
        if (tail[0] === '[]') {
            // the key targets array values so we only traverse if the value is actually an array
            if (Array.isArray(valueAtKey)) {
                for (const item of valueAtKey as any[]) {
                    replacePathsIfSet(item, [tail.slice(1).join('.')], replacement)
                }
            }
        } else if (tail?.length > 0) {
            // we need to traverse the object deeper
            replacePathsIfSet(valueAtKey as any, [tail.join('.')], replacement)
        } else if (valueAtKey !== undefined && valueAtKey !== null) {
            console.log(valueAtKey)
            // we found the key
            ;(obj as any)[head] = replacement
        }
    }
    return obj
}
