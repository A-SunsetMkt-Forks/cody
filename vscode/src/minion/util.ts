export function extractXML(text: string, tag: string): string | null {
    const startTag = `<${tag}>`
    const endTag = `</${tag}>`
    const startIndex = text.indexOf(startTag)
    const endIndex = text.indexOf(endTag)
    if (startIndex === -1 || endIndex === -1) {
        return null
    }
    return text.slice(startIndex + startTag.length, endIndex).trim()
}
