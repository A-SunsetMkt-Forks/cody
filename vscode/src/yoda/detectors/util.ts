import type { CompletionGeneratorValue } from '@sourcegraph/cody-shared'

export function reversedTuple<T extends readonly unknown[]>(tuple: T): T {
    return [...tuple].reverse() as unknown as T
}

export async function combineStream(
    stream: AsyncGenerator<CompletionGeneratorValue>,
    abort?: AbortSignal
) {
    let response = ''
    for await (const message of stream) {
        if (abort?.aborted) {
            return null
        }
        switch (message.type) {
            case 'change':
                response = message.text
                break
            case 'error':
                throw message.error
        }
    }
    if (!response) {
        return null
    }
    return response
}
