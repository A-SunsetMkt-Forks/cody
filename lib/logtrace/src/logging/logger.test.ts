import { afterAll, beforeEach, describe, expect, it } from 'vitest'
import { SaveLogItemsSink, alert, debug, info, log, logger, panic, warn } from '.'
import { dateGenerator, idGenerator } from '../util'

interface TestContext {
    sink: SaveLogItemsSink
}

describe('logger', () => {
    it('works', () => {})
    beforeEach<TestContext>(context => {
        context.sink = reset()
    })
    afterAll(() => {
        idGenerator.setGenerator()
        dateGenerator.setDateFn()
    })
    it<TestContext>('logs basic messages', context => {
        log(debug`This is a message without data`)
        log(info`This is a message with message data ${3}`, { publicExcept: ['msg.0'] })
        log(warn`This is a message with only data`, { value: 3 }, { publicExcept: ['data.value'] })
        log(
            alert`This is a message with both ${'message'} data and data`,
            { value: 5 },
            { publicExcept: ['data.value', 'msg.0'] }
        )

        class User {
            constructor(public readonly name: string) {}
            toJSON() {
                return { jsonName: `json:${this.name}` }
            }
        }
        const user = new User('Bert')
        log(
            panic`This is a complex message ${user.name} ${3}`,
            {
                user, //handles toJSON objects
                array: [{ sheldon: 3 }, { cooper: 'hello' }, { bleh: [{ a: 6 }] }], // handles heterogenous arrays
            },
            { privateExcept: ['data.array', 'msg.0', 'data.user.jsonName', 'data.array.[].bleh.[].a'] }
        )
        expect(context.sink.savedInputs).toMatchSnapshot()
    })
    // it<TestContext>('logs messages to sinks according to log level', context => {
    //     const sink = new TestLogSink()
    //     logger.sinks.add(sink)
    //     logger.minMessageLogLevel = 'info'
    //     const data = {
    //         pub: pub('public'),
    //         priv: priv('private'),
    //         leak: leak('leak'),
    //     }
    //     alert(data)`Public {{pub}} and Private: {{priv}`
    //     warn(data)`Public {{pub}} and Private: {{priv}`
    //     info(data)`Public {{pub}} and Private: {{priv}`
    //     debug(data)`Public {{pub}} and Private: {{priv}}`
    //     expect(sink.items).toMatchSnapshot()
    //     expect(sink.items).toEqual(context.sink.items)
    // })
    // it<TestContext>('infers callsite information', context => {
    //     logger.callStackInference = true
    //     alert`this logs from the root function`
    //     function deepLogCall() {
    //         alert(null, { callsite: 2 })`This logs from within a function but is "invisible"`
    //     }
    //     deepLogCall()
    //     expect(context.sink.items).toMatchSnapshot()
    // })
    // it<TestContext>('can convert errors into object', context => {
    //     logger.callStackInference = true
    //     const err = new Error('Test error')
    //     alert({
    //         err: priv(jsonErr(err)),
    //     })`An error like <{{err.message}}> will automatically get converted into a JSON safe message`
    //     console.log(JSON.stringify(context.sink.items, null, 2))
    // })
    // it<TestContext>('prevents issues through typechecks', context => {
    //     //@ts-expect-error No overload matches this call.
    //     alert({ foo: 'bar' }) //this isn't allowed as you must specify pub(...) or priv(...)
    // })
    // test.todo('can safely leak sensitive information')
    // test.todo('can handle errors with nested causes')
    // test.todo('can tag tests')
    // test.todo('dedents messages')
})

function reset() {
    let seed = 0
    const ts = new Date(Date.UTC(2024, 4, 1))
    const prng = () => {
        seed = (seed * 1103515245 + 12345) % 2 ** 31
        return Math.floor(seed / 65536) % 32768
    }
    idGenerator.setGenerator(prng, 1337)
    dateGenerator.setDateFn(() => {
        return ts
    }, 0)
    const sink = new SaveLogItemsSink()
    logger.register('test', [sink], 'test')
    return sink
}
