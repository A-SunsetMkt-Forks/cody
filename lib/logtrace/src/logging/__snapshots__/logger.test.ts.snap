// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`logger > logs basic messages 1`] = `
[
  {
    "_type": "MessageLogItem",
    "data": {},
    "id": "000000019S0",
    "level": "alert",
    "message": {
      "level": "debug",
      "template": [
        "This is a message without data",
      ],
      "vars": [],
    },
    "origin": {
      "callsite": {
        "column": 9,
        "fullFilePath": "/Users/rnauta/Documents/github/sourcegraph/cody/lib/logtrace/src/logging/logger.test.ts",
        "line": 19,
        "methodName": "<unknown>",
        "relativeFilePath": undefined,
      },
      "id": "test",
      "session": "test",
    },
    "tags": undefined,
    "timestamp": {
      "datetime": 2024-05-01T00:00:00.000Z,
      "timezoneOffset": 0,
    },
    "trace": undefined,
    "verbose": undefined,
    "visibility": {
      "privateExcept": [],
    },
  },
  {
    "_type": "MessageLogItem",
    "data": {},
    "id": "000000019S1",
    "level": "alert",
    "message": {
      "level": "info",
      "template": [
        "This is a message with message data ",
        "",
      ],
      "vars": [
        3,
      ],
    },
    "origin": {
      "callsite": {
        "column": 9,
        "fullFilePath": "/Users/rnauta/Documents/github/sourcegraph/cody/lib/logtrace/src/logging/logger.test.ts",
        "line": 20,
        "methodName": "<unknown>",
        "relativeFilePath": undefined,
      },
      "id": "test",
      "session": "test",
    },
    "tags": undefined,
    "timestamp": {
      "datetime": 2024-05-01T00:00:00.000Z,
      "timezoneOffset": 0,
    },
    "trace": undefined,
    "verbose": undefined,
    "visibility": {
      "privateExcept": [],
    },
  },
  {
    "_type": "MessageLogItem",
    "data": {
      "value": 3,
    },
    "id": "000000019S2",
    "level": "alert",
    "message": {
      "level": "warn",
      "template": [
        "This is a message with only data",
      ],
      "vars": [],
    },
    "origin": {
      "callsite": {
        "column": 9,
        "fullFilePath": "/Users/rnauta/Documents/github/sourcegraph/cody/lib/logtrace/src/logging/logger.test.ts",
        "line": 21,
        "methodName": "<unknown>",
        "relativeFilePath": undefined,
      },
      "id": "test",
      "session": "test",
    },
    "tags": undefined,
    "timestamp": {
      "datetime": 2024-05-01T00:00:00.000Z,
      "timezoneOffset": 0,
    },
    "trace": undefined,
    "verbose": undefined,
    "visibility": {
      "privateExcept": [],
    },
  },
  {
    "_type": "MessageLogItem",
    "data": {
      "value": 5,
    },
    "id": "000000019S3",
    "level": "alert",
    "message": {
      "level": "alert",
      "template": [
        "This is a message with both ",
        " data and data",
      ],
      "vars": [
        "message",
      ],
    },
    "origin": {
      "callsite": {
        "column": 9,
        "fullFilePath": "/Users/rnauta/Documents/github/sourcegraph/cody/lib/logtrace/src/logging/logger.test.ts",
        "line": 22,
        "methodName": "<unknown>",
        "relativeFilePath": undefined,
      },
      "id": "test",
      "session": "test",
    },
    "tags": undefined,
    "timestamp": {
      "datetime": 2024-05-01T00:00:00.000Z,
      "timezoneOffset": 0,
    },
    "trace": undefined,
    "verbose": undefined,
    "visibility": {
      "privateExcept": [],
    },
  },
  {
    "_type": "MessageLogItem",
    "data": {
      "array": [
        {
          "sheldon": 3,
        },
        {
          "cooper": "hello",
        },
      ],
      "user": {
        "jsonName": "json:Bert",
      },
    },
    "id": "000000019S4",
    "level": "alert",
    "message": {
      "level": "panic",
      "template": [
        "This is a complex message ",
        "",
      ],
      "vars": [
        "Bert",
      ],
    },
    "origin": {
      "callsite": {
        "column": 9,
        "fullFilePath": "/Users/rnauta/Documents/github/sourcegraph/cody/lib/logtrace/src/logging/logger.test.ts",
        "line": 35,
        "methodName": "<unknown>",
        "relativeFilePath": undefined,
      },
      "id": "test",
      "session": "test",
    },
    "tags": undefined,
    "timestamp": {
      "datetime": 2024-05-01T00:00:00.000Z,
      "timezoneOffset": 0,
    },
    "trace": undefined,
    "verbose": undefined,
    "visibility": {
      "privateExcept": [],
    },
  },
]
`;
